File: nim_core_v2/android/src/main/kotlin/com/netease/nimflutter/services/FLTFriendService.kt
/*
 * Copyright (c) 2022 NetEase, Inc. All rights reserved.
 * Use of this source code is governed by a MIT license that can be
 * found in the LICENSE file.
 */

package com.netease.nimflutter.services

import android.content.Context
import com.netease.nimflutter.FLTService
import com.netease.nimflutter.LocalError.paramErrorCode
import com.netease.nimflutter.NimCore
import com.netease.nimflutter.NimResult
import com.netease.nimflutter.toMap
import com.netease.nimflutter.toV2NIMFriendAddApplication
import com.netease.nimlib.sdk.NIMClient
import com.netease.nimlib.sdk.v2.friend.V2NIMFriend
import com.netease.nimlib.sdk.v2.friend.V2NIMFriendAddApplication
import com.netease.nimlib.sdk.v2.friend.V2NIMFriendListener
import com.netease.nimlib.sdk.v2.friend.V2NIMFriendService
import com.netease.nimlib.sdk.v2.friend.enums.V2NIMFriendAddApplicationStatus
import com.netease.nimlib.sdk.v2.friend.enums.V2NIMFriendAddMode
import com.netease.nimlib.sdk.v2.friend.enums.V2NIMFriendDeletionType
import com.netease.nimlib.sdk.v2.friend.option.V2NIMFriendAddApplicationQueryOption.V2NIMFriendAddApplicationQueryOptionBuilder
import com.netease.nimlib.sdk.v2.friend.option.V2NIMFriendSearchOption
import com.netease.nimlib.sdk.v2.friend.param.V2NIMFriendAddParams
import com.netease.nimlib.sdk.v2.friend.param.V2NIMFriendDeleteParams
import com.netease.nimlib.sdk.v2.friend.param.V2NIMFriendSetParams
import com.netease.yunxin.kit.alog.ALog
import kotlin.coroutines.resume
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.channels.onFailure
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.suspendCancellableCoroutine

/**
 * Flutter 好友服务
 *
 */
class FLTFriendService(
    applicationContext: Context,
    nimCore: NimCore
) : FLTService(applicationContext, nimCore) {
    override val serviceName: String = "FriendService"

    init {
        nimCore.onInitialized {
            friendListener()
            registerFlutterMethodCalls(
                "addFriend" to this::addFriend,
                "deleteFriend" to this::deleteFriend,
                "acceptAddApplication" to this::acceptAddApplication,
                "rejectAddApplication" to this::rejectAddApplication,
                "setFriendInfo" to this::setFriendInfo,
                "getFriendList" to this::getFriendList,
                "getFriendByIds" to this::getFriendByIds,
                "checkFriend" to this::checkFriend,
                "getAddApplicationList" to this::getAddApplicationList,
                "getAddApplicationUnreadCount" to this::getAddApplicationUnreadCount,
                "setAddApplicationRead" to this::setAddApplicationRead,
                "searchFriendByOption" to this::searchFriendByOption
            )
        }
    }

    @ExperimentalCoroutinesApi
    private fun friendListener() {
        callbackFlow<Pair<String, Map<String, Any?>?>> {
            val listener = object : V2NIMFriendListener {
                override fun onFriendAdded(friendInfo: V2NIMFriend?) {
                    ALog.i(serviceName, "onFriendAdded: $friendInfo")
                    trySend(
                        Pair<String, Map<String, Any?>?>(
                            "onFriendAdded",
                            friendInfo?.toMap()
                        )
                    ).onFailure {
                        ALog.i(serviceName, "send onFriendAdded fail: ${it?.message}")
                    }
                }

                override fun onFriendDeleted(
                    accountId: String?,
                    deletionType: V2NIMFriendDeletionType?
                ) {
                    ALog.i(serviceName, "onFriendDeleted: $accountId")
                    trySend(
                        Pair<String, Map<String, Any?>?>(
                            "onFriendDeleted",
                            mapOf(
                                "accountId" to accountId,
                                "deletionType" to deletionType?.value
                            )
                        )
                    ).onFailure {
                        ALog.i(serviceName, "send onFriendDeleted fail: ${it?.message}")
                    }
                }

                override fun onFriendAddApplication(applicationInfo: V2NIMFriendAddApplication?) {
                    ALog.i(serviceName, "onFriendAddApplication: ${applicationInfo?.applicantAccountId}")
                    trySend(
                        Pair<String, Map<String, Any?>?>(
                            "onFriendAddApplication",
                            applicationInfo?.toMap()
                        )
                    ).onFailure {
                        ALog.i(serviceName, "send onFriendAddApplication fail: ${it?.message}")
                    }
                }

                override fun onFriendAddRejected(rejectionInfo: V2NIMFriendAddApplication?) {
                    ALog.i(serviceName, "onFriendAddRejected: $rejectionInfo")
                    trySend(
                        Pair<String, Map<String, Any?>?>(
                            "onFriendAddRejected",
                            rejectionInfo?.toMap()
                        )
                    ).onFailure {
                        ALog.i(serviceName, "send onFriendAddRejected fail: ${it?.message}")
                    }
                }

                override fun onFriendInfoChanged(friendInfo: V2NIMFriend?) {
                    ALog.i(serviceName, "onFriendInfoChanged: $friendInfo")
                    trySend(
                        Pair<String, Map<String, Any?>?>(
                            "onFriendInfoChanged",
                            friendInfo?.toMap()
                        )
                    ).onFailure {
                        ALog.i(serviceName, "send onFriendInfoChanged fail: ${it?.message}")
                    }
                }
            }
            NIMClient.getService(V2NIMFriendService::class.java).apply {
                this.addFriendListener(listener)
                awaitClose {
                    this.removeFriendListener(listener)
                }
            }
        }.onEach { event ->
            notifyEvent(
                method = event.first,
                arguments = event.second as Map<String, Any?>
            )
        }.launchIn(nimCore.lifeCycleScope)
    }

    private suspend fun addFriend(arguments: Map<String, *>): NimResult<Void> {
        val accountId = arguments["accountId"] as? String
        val params = arguments["params"] as? Map<String, *>
        return suspendCancellableCoroutine { cont ->
            var addParams: V2NIMFriendAddParams? = null
            if (params?.isNotEmpty() == true) {
                val mode = V2NIMFriendAddMode.typeOfValue((params["addMode"] as? Int ?: 1).toByte())
                addParams = V2NIMFriendAddParams.V2NIMFriendAddParamsBuilder.builder(
                    mode
                ).withPostscript(params["postscript"] as? String)
                    .build()
            }
            NIMClient.getService(V2NIMFriendService::class.java).addFriend(
                accountId,
                addParams,
                {
                    cont.resume(NimResult(0, data = it))
                },
                {
                    cont.resume(NimResult(it.code, errorDetails = it.desc))
                }
            )
        }
    }

    private suspend fun deleteFriend(arguments: Map<String, *>): NimResult<Void> {
        val accountId = arguments["accountId"] as? String
        val params = arguments["params"] as? Map<String, *>
        if (accountId?.isEmpty() == true) {
            return NimResult(code = paramErrorCode, errorDetails = "accountId is empty")
        }
        return suspendCancellableCoroutine { cont ->
            var deleteParams: V2NIMFriendDeleteParams? = null
            if (params?.isNotEmpty() == true) {
                deleteParams = V2NIMFriendDeleteParams.V2NIMFriendDeleteParamsBuilder.builder().withDeleteAlias(params["deleteAlias"] as? Boolean ?: false)
                    .build()
            }
            NIMClient.getService(V2NIMFriendService::class.java).deleteFriend(
                accountId,
                deleteParams,
                {
                    cont.resume(NimResult(0, data = it))
                },
                {
                    cont.resume(NimResult(it.code, errorDetails = it.desc))
                }
            )
        }
    }

    private suspend fun acceptAddApplication(arguments: Map<String, *>): NimResult<Void> {
        val applicationMap = arguments["application"] as? Map<String, *>
        val application = applicationMap?.toV2NIMFriendAddApplication()
        ALog.d(serviceName, "acceptAddApplication: ${application?.applicantAccountId}")
        return suspendCancellableCoroutine { cont ->
            NIMClient.getService(V2NIMFriendService::class.java).acceptAddApplication(
                application,
                {
                    cont.resume(NimResult(0, data = it))
                },
                {
                    cont.resume(NimResult(it.code, errorDetails = it.desc))
                }
            )
        }
    }

    private suspend fun rejectAddApplication(arguments: Map<String, *>): NimResult<Void> {
        val postscript = arguments["postscript"] as? String
        val applicationMap = arguments["application"] as? Map<String, *>
        if (applicationMap?.isEmpty() == true) {
            return NimResult(code = paramErrorCode, errorDetails = "application is empty")
        }
        return suspendCancellableCoroutine { cont ->
            NIMClient.getService(V2NIMFriendService::class.java).rejectAddApplication(
                applicationMap!!.toV2NIMFriendAddApplication(),
                postscript,
                {
                    cont.resume(NimResult(0, data = it))
                },
                {
                    cont.resume(NimResult(it.code, errorDetails = it.desc))
                }
            )
        }
    }

    private suspend fun setFriendInfo(arguments: Map<String, *>): NimResult<Void> {
        val accountId = arguments["accountId"] as? String
        val paramsMap = arguments["params"] as? Map<String, *>
//        if (accountId?.isEmpty() == true) {
//            return NimResult(code = paramErrorCode, errorDetails = "accountId is empty")
//        }
        return suspendCancellableCoroutine { cont ->
            var paramsBuilder = V2NIMFriendSetParams.V2NIMFriendSetParamsBuilder.builder()
            if (paramsMap?.isNotEmpty() == true) {
                if (paramsMap.containsKey("alias")) {
                    paramsBuilder = paramsBuilder.withAlias(paramsMap["alias"] as? String)
                }
                if (paramsMap.containsKey("serverExtension")) {
                    paramsBuilder = paramsBuilder.withServerExtension(paramsMap["serverExtension"] as? String)
                }
            }
            val params = paramsBuilder.build()
            NIMClient.getService(V2NIMFriendService::class.java).setFriendInfo(
                accountId,
                params,
                {
                    cont.resume(NimResult(0, data = it))
                },
                {
                    cont.resume(NimResult(it.code, errorDetails = it.desc))
                }
            )
        }
    }

    private suspend fun getFriendList(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            NIMClient.getService(V2NIMFriendService::class.java).getFriendList(
                {
                    cont.resume(
                        NimResult(
                            0,
                            data = mapOf(
                                "friendList" to it.map { friend -> friend.toMap() }.toList()
                            )
                        )
                    )
                },
                {
                    cont.resume(NimResult(it.code, errorDetails = it.desc))
                }
            )
        }
    }

    private suspend fun getFriendByIds(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        val accountIds = arguments["accountIds"] as? List<String>
//        if (accountIds?.isEmpty() == true) {
//            return NimResult(code = paramErrorCode, errorDetails = "accountIds is empty")
//        }
        return suspendCancellableCoroutine { cont ->
            NIMClient.getService(V2NIMFriendService::class.java).getFriendByIds(
                accountIds,
                {
                    cont.resume(
                        NimResult(
                            0,
                            data = mapOf(
                                "friendList" to it.map { friend -> friend.toMap() }.toList()
                            )
                        )
                    )
                },
                {
                    cont.resume(NimResult(it.code, errorDetails = it.desc))
                }
            )
        }
    }

    private suspend fun checkFriend(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        val accountIds = arguments["accountIds"] as? List<String>
        return suspendCancellableCoroutine { cont ->
            NIMClient.getService(V2NIMFriendService::class.java).checkFriend(
                accountIds,
                {
                    cont.resume(NimResult(0, data = mapOf("result" to it)))
                },
                {
                    cont.resume(NimResult(it.code, errorDetails = it.desc))
                }
            )
        }
    }

    private suspend fun getAddApplicationList(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        var optionMap = arguments["option"] as? Map<String, *>
        if (optionMap?.isEmpty() == true) {
            return NimResult(code = paramErrorCode, errorDetails = "option is empty")
        }
        return suspendCancellableCoroutine { cont ->
            val option = V2NIMFriendAddApplicationQueryOptionBuilder.builder().withLimit(
                optionMap!!["limit"] as? Int ?: 50
            ).withOffset(
                (optionMap["offset"] as? Int ?: 0).toLong()
            ).withStatus(
                (optionMap["status"] as? List<Int>)?.map {
                        status ->
                    V2NIMFriendAddApplicationStatus.typeOfValue(status)
                }?.toList() ?: emptyList()
            ).build()
            NIMClient.getService(V2NIMFriendService::class.java).getAddApplicationList(
                option,
                {
                    cont.resume(
                        NimResult(
                            0,
                            data = mapOf(
                                "infos" to it.infos.map { application -> application.toMap() }.toList(),
                                "offset" to it.offset,
                                "finished" to it.isFinished
                            )
                        )
                    )
                },
                {
                    cont.resume(NimResult(it.code, errorDetails = it.desc))
                }
            )
        }
    }

    private suspend fun getAddApplicationUnreadCount(arguments: Map<String, *>): NimResult<Int> {
        return suspendCancellableCoroutine { cont ->
            NIMClient.getService(V2NIMFriendService::class.java).getAddApplicationUnreadCount(
                {
                    cont.resume(NimResult(0, data = it))
                },
                {
                    cont.resume(NimResult(it.code, errorDetails = it.desc))
                }
            )
        }
    }

    private suspend fun setAddApplicationRead(arguments: Map<String, *>): NimResult<Void> {
        return suspendCancellableCoroutine { cont ->
            NIMClient.getService(V2NIMFriendService::class.java).setAddApplicationRead(
                {
                    cont.resume(NimResult(0, data = it))
                },
                {
                    cont.resume(NimResult(it.code, errorDetails = it.desc))
                }
            )
        }
    }

    private suspend fun searchFriendByOption(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        val optionMap = arguments["friendSearchOption"] as? Map<String, *>
        if (optionMap?.isEmpty() == true) {
            return NimResult(code = paramErrorCode, errorDetails = "option is empty")
        }
        return suspendCancellableCoroutine { cont ->
            val option = V2NIMFriendSearchOption.V2NIMFriendSearchOptionBuilder.builder(
                optionMap!!["keyword"] as? String ?: ""
            ).withSearchAlias(
                optionMap["searchAlias"] as? Boolean ?: true
            ).withSearchAccountId(
                optionMap["searchAccountId"] as? Boolean ?: true
            ).build()
            NIMClient.getService(V2NIMFriendService::class.java).searchFriendByOption(
                option,
                {
                    cont.resume(
                        NimResult(
                            0,
                            data = mapOf(
                                "friendList" to it.map { friend -> friend.toMap() }.toList()
                            )
                        )
                    )
                },
                {
                    cont.resume(NimResult(it.code, errorDetails = it.desc))
                }
            )
        }
    }
}

File: nim_core_v2/ios/Classes/services/FLTFriendService.swift
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import Foundation
import NIMSDK

class FLTFriendService: FLTBaseService, FLTService,
  V2NIMFriendListener {
  func serviceName() -> String {
    ServiceType.FriendService.rawValue
  }

  private let paramErrorTip = "param Error"
  private let paramErrorCode = 199_414

  private func friendCallback(_ error: Error?, _ data: Any?, _ resultCallback: ResultCallback) {
    if let ns_error = error as NSError? {
      errorCallBack(resultCallback, ns_error.description, ns_error.code)
    } else {
      successCallBack(resultCallback, data)
    }
  }

  override func onInitialized() {
    NIMSDK.shared().v2FriendService.add(self)
  }

  deinit {
    NIMSDK.shared().v2FriendService.remove(self)
  }

  func onMethodCalled(_ method: String, _ arguments: [String: Any], _ resultCallback: ResultCallback) {
    switch method {
    case "addFriend":
      addFriend(arguments, resultCallback)
    case "deleteFriend":
      deleteFriend(arguments, resultCallback)
    case "setFriendInfo":
      setFriendInfo(arguments, resultCallback)
    case "acceptAddApplication":
      acceptAddApplication(arguments, resultCallback)
    case "rejectAddApplication":
      rejectAddApplication(arguments, resultCallback)
    case "getFriendList":
      getFriendList(arguments, resultCallback)
    case "getFriendByIds":
      getFriendByIds(arguments, resultCallback)
    case "searchFriendByOption":
      searchFriendByOption(arguments, resultCallback)
    case "checkFriend":
      checkFriend(arguments, resultCallback)
    case "getAddApplicationList":
      getAddApplicationList(arguments, resultCallback)
    case "getAddApplicationUnreadCount":
      getAddApplicationUnreadCount(arguments, resultCallback)
    case "setAddApplicationRead":
      setAddApplicationRead(arguments, resultCallback)
    case "clearAllAddApplication":
      clearAllAddApplication(arguments, resultCallback)
    case "clearAllAddApplicationEx":
      clearAllAddApplicationEx(arguments, resultCallback)
    case "deleteAddApplication":
      deleteAddApplication(arguments, resultCallback)
    default:
      resultCallback.notImplemented()
    }
  }

  func register(_ nimCore: NimCore) {
    self.nimCore = nimCore
    nimCore.addService(self)
  }

  func onFriendAdded(_ friendInfo: V2NIMFriend) {
    notifyEvent(serviceName(), "onFriendAdded", friendInfo.toDic() as [String: Any])
  }

  func onFriendDeleted(_ accountId: String, deletionType: V2NIMFriendDeletionType) {
    notifyEvent(serviceName(), "onFriendDeleted", ["accountId": accountId, "deletionType": deletionType.rawValue])
  }

  func onFriendAddApplication(_ application: V2NIMFriendAddApplication) {
    notifyEvent(serviceName(), "onFriendAddApplication", application.toDic())
  }

  func onFriendAddRejected(_ rejectionInfo: V2NIMFriendAddApplication) {
    notifyEvent(serviceName(), "onFriendAddRejected", rejectionInfo.toDic())
  }

  func onFriendInfoChanged(_ friendInfo: V2NIMFriend) {
    if let user = friendInfo.userProfile {
      notifyEvent(serviceName(), "onFriendInfoChanged", friendInfo.toDic())
    } else {
      weak var weakSelf = self
      NIMSDK.shared().v2UserService.getUserList([friendInfo.accountId!]) { users in
        if !users.isEmpty {
          friendInfo.setValue(users[0], forKeyPath: #keyPath(V2NIMFriend.userProfile))
        }
        weakSelf?.notifyEvent(weakSelf?.serviceName() ?? "FriendService", "onFriendInfoChanged", friendInfo.toDic())
      }
    }
  }

  private func addFriend(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let accountId = arguments["accountId"] as? String else {
      errorCallBack(resultCallback, paramErrorTip, paramErrorCode)
      return
    }

    var params = V2NIMFriendAddParams()
    if let paramsMap = arguments["params"] as? [String: Any] {
      params = V2NIMFriendAddParams.fromDic(paramsMap)
    }

    weak var weakSelf = self
    NIMSDK.shared().v2FriendService.addFriend(accountId, params: params) {
      weakSelf?.friendCallback(nil, nil, resultCallback)
    } failure: { error in
      weakSelf?.friendCallback(error.nserror, nil, resultCallback)
    }
  }

  private func deleteFriend(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let accountId = arguments["accountId"] as? String else {
      errorCallBack(resultCallback, paramErrorTip, paramErrorCode)
      return
    }

    var params = V2NIMFriendDeleteParams()
    if let paramsMap = arguments["params"] as? [String: Any] {
      params = V2NIMFriendDeleteParams.fromDic(paramsMap)
    }

    weak var weakSelf = self
    NIMSDK.shared().v2FriendService.deleteFriend(accountId, params: params) {
      weakSelf?.friendCallback(nil, nil, resultCallback)
    } failure: { error in
      weakSelf?.friendCallback(error.nserror, nil, resultCallback)
    }
  }

  private func setFriendInfo(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let accountId = arguments["accountId"] as? String else {
      errorCallBack(resultCallback, paramErrorTip, paramErrorCode)
      return
    }

    var params = V2NIMFriendSetParams()
    if let paramsMap = arguments["params"] as? [String: Any] {
      params = V2NIMFriendSetParams.fromDic(paramsMap)
    }

    weak var weakSelf = self
    NIMSDK.shared().v2FriendService.setFriendInfo(accountId, params: params) {
      weakSelf?.friendCallback(nil, nil, resultCallback)
    } failure: { error in
      weakSelf?.friendCallback(error.nserror, nil, resultCallback)
    }
  }

  private func acceptAddApplication(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    var application = V2NIMFriendAddApplication()
    if let paramsMap = arguments["application"] as? [String: Any] {
      application = V2NIMFriendAddApplication.fromDic(paramsMap)
    }

    weak var weakSelf = self
    NIMSDK.shared().v2FriendService.accept(application) {
      weakSelf?.friendCallback(nil, nil, resultCallback)
    } failure: { error in
      weakSelf?.friendCallback(error.nserror, nil, resultCallback)
    }
  }

  private func rejectAddApplication(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    var application = V2NIMFriendAddApplication()
    if let paramsMap = arguments["application"] as? [String: Any] {
      application = V2NIMFriendAddApplication.fromDic(paramsMap)
    }

    let postscript = arguments["postscript"] as? String ?? ""

    weak var weakSelf = self
    NIMSDK.shared().v2FriendService.reject(application, postscript: postscript) {
      weakSelf?.friendCallback(nil, nil, resultCallback)
    } failure: { error in
      weakSelf?.friendCallback(error.nserror, nil, resultCallback)
    }
  }

  private func getFriendList(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    weak var weakSelf = self
    NIMSDK.shared().v2FriendService.getFriendList { friends in
      weakSelf?.friendCallback(nil,
                               ["friendList": friends.map { $0.toDic() }],
                               resultCallback)
    } failure: { error in
      weakSelf?.friendCallback(error.nserror, nil, resultCallback)
    }
  }

  private func getFriendByIds(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let accountIds = arguments["accountIds"] as? [String] else {
      errorCallBack(resultCallback, paramErrorTip, paramErrorCode)
      return
    }

    weak var weakSelf = self
    NIMSDK.shared().v2FriendService.getFriendByIds(accountIds) { friends in
      weakSelf?.friendCallback(nil,
                               ["friendList": friends.map { $0.toDic() }],
                               resultCallback)
    } failure: { error in
      weakSelf?.friendCallback(error.nserror, nil, resultCallback)
    }
  }

  private func searchFriendByOption(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let optionMap = arguments["friendSearchOption"] as? [String: Any] else {
      errorCallBack(resultCallback, paramErrorTip, paramErrorCode)
      return
    }

    let option = V2NIMFriendSearchOption.fromDic(optionMap)

    weak var weakSelf = self
    NIMSDK.shared().v2FriendService.searchFriend(by: option) { friends in
      weakSelf?.friendCallback(nil,
                               ["friendList": friends.map { $0.toDic() }],
                               resultCallback)
    } failure: { error in
      weakSelf?.friendCallback(error.nserror, nil, resultCallback)
    }
  }

  private func checkFriend(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let accountIds = arguments["accountIds"] as? [String] else {
      errorCallBack(resultCallback, paramErrorTip, paramErrorCode)
      return
    }

    weak var weakSelf = self
    NIMSDK.shared().v2FriendService.checkFriend(accountIds) { result in
      weakSelf?.friendCallback(nil, ["result": result], resultCallback)
    } failure: { error in
      weakSelf?.friendCallback(error.nserror, nil, resultCallback)
    }
  }

  private func getAddApplicationList(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    var option = V2NIMFriendAddApplicationQueryOption()
    if let optionMap = arguments["option"] as? [String: Any] {
      option = V2NIMFriendAddApplicationQueryOption.fromDic(optionMap)
    }

    weak var weakSelf = self
    NIMSDK.shared().v2FriendService.getAddApplicationList(option) { result in
      weakSelf?.friendCallback(nil,
                               result.toDic(),
                               resultCallback)
    } failure: { error in
      weakSelf?.friendCallback(error.nserror, nil, resultCallback)
    }
  }

  private func getAddApplicationUnreadCount(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    weak var weakSelf = self
    NIMSDK.shared().v2FriendService.getAddApplicationUnreadCount { count in
      weakSelf?.friendCallback(nil, count, resultCallback)
    } failure: { error in
      weakSelf?.friendCallback(error.nserror, nil, resultCallback)
    }
  }

  private func setAddApplicationRead(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    weak var weakSelf = self
    NIMSDK.shared().v2FriendService.setAddApplicationRead { _ in
      weakSelf?.friendCallback(nil, nil, resultCallback)
    } failure: { error in
      weakSelf?.friendCallback(error.nserror, nil, resultCallback)
    }
  }

  private func clearAllAddApplication(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    weak var weakSelf = self
    NIMSDK.shared().v2FriendService.clearAllAddApplication {
      weakSelf?.friendCallback(nil, nil, resultCallback)
    } failure: { error in
      weakSelf?.friendCallback(error.nserror, nil, resultCallback)
    }
  }

  private func clearAllAddApplicationEx(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    let option = V2NIMFriendClearAddApplicationOption()
    if let optionMap = arguments["option"] as? [String: Any] {
      if let timestamp = optionMap["timestamp"] as? NSNumber {
        option.timestamp = timestamp.doubleValue / 1000.0 // Dart uses milliseconds, iOS uses seconds
      }
      if let typeRaw = optionMap["type"] as? Int,
         let type = V2NIMFriendAddApplicationType(rawValue: typeRaw) {
        // Dart: 0=legacy, 1=fromSelf, 2=toSelf, 3=all → iOS rawValue 与 Android 对齐
        option.type = type
      }
    }
    weak var weakSelf = self
    NIMSDK.shared().v2FriendService.clearAllAddApplicationEx(option) {
      weakSelf?.friendCallback(nil, nil, resultCallback)
    } failure: { error in
      weakSelf?.friendCallback(error.nserror, nil, resultCallback)
    }
  }

  private func deleteAddApplication(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    var application = V2NIMFriendAddApplication()
    if let paramsMap = arguments["application"] as? [String: Any] {
      application = V2NIMFriendAddApplication.fromDic(paramsMap)
    }

    weak var weakSelf = self
    NIMSDK.shared().v2FriendService.delete(application) {
      weakSelf?.friendCallback(nil, nil, resultCallback)
    } failure: { error in
      weakSelf?.friendCallback(error.nserror, nil, resultCallback)
    }
  }
}
